presto-main/src/main/java/com/facebook/presto/dispatcher/QueryInfoFetcher.java

```diff
@@ -0,0 +1,323 @@
+        this.updateScheduledExecutor = requireNonNull(updateScheduledExecutor, "updateScheduledExecutor is null");
+        this.errorScheduledExecutor = requireNonNull(errorScheduledExecutor, "errorScheduledExecutor is null");
+        this.remoteQueryStats = requireNonNull(remoteQueryStats, "remoteQueryStats is null");
+        this.httpLocationFactory = requireNonNull(httpLocationFactory, "httpLocationFactory is null");
+        this.isBinaryTransportEnabled = isBinaryTransportEnabled;
+    }
+
+    public BasicQueryInfo getQueryInfo()
+    {
+        return queryInfo.get();
+    }
+
+    public synchronized void start(InternalNode coordinator)
+    {
+        checkState(!running, "Already running");
+        checkState(!stopped, "Already stopped");
+        requireNonNull(coordinator, "coordinator is null");
+        checkArgument(coordinator.isCoordinator(), "node %s is not a coordinator", coordinator.getInternalUri());
+        URI location = httpLocationFactory.createQueryLocation(coordinator, queryId);
+        location = UriBuilder.fromUri(location).queryParam("basic", true).build();
+        this.errorTracker = queryRequestErrorTracker(queryId, location, maxErrorDuration, errorScheduledExecutor);
+        scheduleUpdate(location);
+        running = true;
+    }
+
+    synchronized void stop()
+    {
+        running = false;
+        stopped = true;
+        if (future != null) {
+            // do not terminate if the request is already running to avoid closing pooled connections
+            future.cancel(false);
+            future = null;
+        }
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel(true);
+            scheduledFuture = null;
+        }
+    }
+
+    public void addFinalQueryInfoListener(StateMachine.StateChangeListener<BasicQueryInfo> stateChangeListener)
+    {
+        AtomicBoolean done = new AtomicBoolean();
+        StateMachine.StateChangeListener<Optional<BasicQueryInfo>> fireOnceStateChangeListener = finalQueryInfo -> {
+            if (finalQueryInfo.isPresent() && done.compareAndSet(false, true)) {
+                stateChangeListener.stateChanged(finalQueryInfo.get());
+            }
+        };
+        finalQueryInfo.addStateChangeListener(fireOnceStateChangeListener);
+        fireOnceStateChangeListener.stateChanged(finalQueryInfo.get());
+    }
+
+    private synchronized void scheduleUpdate(URI location)
+    {
+        checkState(!running, "Already scheduled");
+        scheduledFuture = updateScheduledExecutor.scheduleWithFixedDelay(() -> {
+            synchronized (this) {
+                // if the previous request still running, don't schedule a new request
+                if (future != null && !future.isDone()) {
+                    return;
+                }
+            }
+            if (nanosSince(lastUpdateNanos.get()).toMillis() >= updateIntervalMillis) {
+                sendNextRequest(location);
```

Comment:
Please try-catch it, in case it ever fails. Remember that if an exception is thrown out of the `scheduleWithFixedDelay` the scheduled task dies